File: 106. Construct Binary Tree from Inorder and Postorder Traversal/106. Construct Binary Tree from Inorder and Postorder Traversal.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
def build_tree_helper(inorder, in_start, in_end, postorder, post_start, post_end, in_map)
    return nil if in_start > in_end
    
    root = TreeNode.new(postorder[post_end])
    root_index = in_map[root.val]
    left_subtree_size = root_index - in_start
    
    root.left = build_tree_helper(inorder, in_start, root_index - 1,
                                 postorder, post_start, post_start + left_subtree_size - 1,
                                 in_map)
    root.right = build_tree_helper(inorder, root_index + 1, in_end,
                                  postorder, post_start + left_subtree_size, post_end - 1,
                                  in_map)
    
    root
end

def build_tree(inorder, postorder)
    in_map = {}
    inorder.each_with_index { |val, idx| in_map[val] = idx }
    
    build_tree_helper(inorder, 0, inorder.length - 1,
                     postorder, 0, postorder.length - 1,
                     in_map)
end

File: 106. Construct Binary Tree from Inorder and Postorder Traversal/106. Construct Binary Tree from Inorder and Postorder Traversal.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func buildTreeHelper(inorder []int, inStart, inEnd int,
	postorder []int, postStart, postEnd int,
	inMap map[int]int) *TreeNode {
if inStart > inEnd {
return nil
}

root := &TreeNode{Val: postorder[postEnd]}
rootIndex := inMap[root.Val]
leftSubtreeSize := rootIndex - inStart

root.Left = buildTreeHelper(inorder, inStart, rootIndex - 1,
			   postorder, postStart, postStart + leftSubtreeSize - 1,
			   inMap)
root.Right = buildTreeHelper(inorder, rootIndex + 1, inEnd,
				postorder, postStart + leftSubtreeSize, postEnd - 1,
				inMap)

return root
}

func buildTree(inorder []int, postorder []int) *TreeNode {
inMap := make(map[int]int)

for i, val := range inorder {
inMap[val] = i
}

return buildTreeHelper(inorder, 0, len(inorder) - 1,
		  postorder, 0, len(postorder) - 1,
		  inMap)
}